### ISSUE TITLE
uri check sometime fails

### ISSUE THREAD

**Issue by xadhoom** [hrzndhrn/json_xema#36]
While testing my code with StreamData, occasionally a string of uri type is reported as not valid, while it respects the format defined in RFC3986. Other parsers does not report any error.

For example, an uri like: `xgQWvkZIXoyytbJQSRuewfioiCTySkvtBVMexp://kfeqBN1XcrQ1p9agSBWOqfZ462EJOI9zBLsGsIREJmYkpxG2JnzqqKF8T2oOrgD4T.com` is reported as wrong, while it is not.

A very basic code example:
```elixir
  property "with valid uri" do
      schema =
        """
        {
          "type": "object",
          "properties": {
            "some_url": {
              "type": "string",
              "format": "uri"
            }
          }
        }
        """
        |> Jason.decode!()
        |> JsonXema.new()

      uri_segment_chars = Enum.concat([?0..?9, ?a..?z, ?A..?Z])
      scheme_segment_chars = Enum.concat([?a..?z, ?A..?Z])

      check all(
              scheme <-
                StreamData.string(scheme_segment_chars, min_length: 1),
              host <- StreamData.string(uri_segment_chars, min_length: 1),
              path <- StreamData.string(uri_segment_chars)
            ) do
        data = %{"some_url" => "#{scheme}://#{host}.com/#{path}"}

        assert :ok = JsonXema.validate(schema, data)
      end
    end
```

Another quick test:
```elixir
iex(20)> Xema.Format.uri?("xgQWvkZIXoyytbJQSRuewfioiCTySkvtBVMexp://kfeqBN1XcrQ1p9agSBWOqfZ462EJOrgDaaaa4aaaaaaaaaaaaaaaaaaaaaaaaaa.com")
true
iex(21)> Xema.Format.uri?("xgQWvkZIXoyytbJQSRuewfioiCTySkvtBVMexp://kfeqBN1XcrQ1p9agSBWOqfZ462EJOrgDaaaa4aaaaaaaaaaaaaaaaaaaaaaaaaaa.com")
false
```

seems related in some way to the string len. According to RFC, while there's no formal limit a limit of 255 is recommended, but this is not the case (we're under 100).

There's a limit of 63 chars for each label under the "Domain name syntax, internationalization" rules, but strictly speaking this is not required by the Json Schema standard nor from the RFC3986.

I suggest to at least document it or just enforce the standard without further processing.

**Comment by NickNeck** [comment]
Thanks for reporting. I will take a look.

**Comment by NickNeck** [comment]
> There's a limit of 63 chars for each label under the "Domain name syntax, internationalization" rules, but strictly speaking this is not required by the Json Schema standard nor from the RFC3986.

In [RFC 1034](https://tools.ietf.org/html/rfc1034) and [RFC 1123](https://tools.ietf.org/html/rfc1123) the length of a subdomain is restricted to 63. But you are right that JsonSchema doesn't demand this for `uri` but just for `hostname`. In `Xema`, the `hostname` validator is also be used for `uri`. I will fix this in the next days. Thanks again.

**Comment by NickNeck** [comment]
Should be fixed with `Xema 0.13.3`.
